core/src/auth_tokens.rs

```diff
@@ -0,0 +1,484 @@
+impl AsRef<str> for DapAuthToken {
+    fn as_ref(&self) -> &str {
+        &self.0
+    }
+}
+
+impl AsRef<[u8]> for DapAuthToken {
+    fn as_ref(&self) -> &[u8] {
+        self.0.as_bytes()
+    }
+}
+
+impl TryFrom<String> for DapAuthToken {
+    type Error = anyhow::Error;
+
+    fn try_from(value: String) -> Result<Self, Self::Error> {
+        Self::validate(&value)?;
+        Ok(Self(value))
+    }
+}
+
+impl TryFrom<Vec<u8>> for DapAuthToken {
+    type Error = anyhow::Error;
+
+    fn try_from(value: Vec<u8>) -> Result<Self, Self::Error> {
+        Self::try_from(String::from_utf8(value)?)
+    }
+}
+
+impl<'de> Deserialize<'de> for DapAuthToken {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        String::deserialize(deserializer)
+            .and_then(|string| Self::try_from(string).map_err(D::Error::custom))
+    }
+}
+
+impl PartialEq for DapAuthToken {
+    fn eq(&self, other: &Self) -> bool {
+        // We attempt constant-time comparisons of the token data to mitigate timing attacks. Note
+        // that this function still leaks whether the lengths of the tokens are equal -- this is
+        // acceptable because we expect the content of the tokens to provide enough randomness that
+        // needs to be guessed even if the length is known.
+        constant_time::verify_slices_are_equal(self.0.as_ref(), other.0.as_ref()).is_ok()
+    }
+}
+
+impl Eq for DapAuthToken {}
+
+impl Distribution<DapAuthToken> for Standard {
+    fn sample<R: rand::Rng + ?Sized>(&self, rng: &mut R) -> DapAuthToken {
+        DapAuthToken(URL_SAFE_NO_PAD.encode(rng.gen::<[u8; 16]>()))
+    }
+}
+
+/// A token value used to authenticate HTTP requests. This token is used in "Authorization: Bearer"
+/// HTTP request headers.
+///
+/// Token values must follow the syntax in
+/// <https://datatracker.ietf.org/doc/html/rfc6750#section-2.1>. Its literal value is the canonical
+/// form of the token and is used directly, without any additional encoding or decoding, in
+/// configuration files, Janus aggregator API requests, and HTTP authentication headers.
+///
+/// This opaque type ensures it's impossible to construct an [`AuthenticationToken`] whose contents
+/// are invalid.
+#[derive(Clone, Derivative, Serialize)]
+#[derivative(Debug)]
+#[serde(transparent)]
+pub struct BearerToken(#[derivative(Debug = "ignore")] String);
+
+impl BearerToken {
+    /// Returns the token as a string.
+    pub fn as_str(&self) -> &str {
+        &self.0
+    }
+
+    /// Validate that a bearer token value matches the format in
+    /// https://datatracker.ietf.org/doc/html/rfc6750#section-2.1.
+    fn validate(value: &str) -> Result<(), anyhow::Error> {
+        let mut iter = value.chars();
+        let mut any_non_equals = false;
+        // First loop: consume "normal" characters, stop when we see an equals sign for padding or
+        // reach the end of the input.
+        for c in &mut iter {
```

Comment:
Any reason not to use a regex here? I lean away from them normally but a small value defined by BNF-like grammar is a perfect place for a regex IMO.

I think it would be something like `[-a-zA-Z0-9._~+/]+=*`.

Reply:
(we could also just try decoding as base64, since I suppose that's the intent -- though that does lose some non-base64 values that are allowed by the BNF, such as `A=======`.)

Reply:
I introduced this in #1978. It got moved in #1984, and it'll disappear from here once it's rebased on main. A regex sounds good, it may be able to run faster too. As described in #1978, base64-decoding would be more strict than necessary, both in terms of the character set allowed, the choice of padding vs. no padding, and limitations on length (excluding padding). I can make a follow-up PR to swap in a regex here.

Reply:
Sorry, I should have called out that this PR is stacked on #1984.